### PR TITLE
Add typechecking for RecursiveRecordTypeMerge

### LINKED ISSUE #91: Implement Type Check for //\\
The implementation of type-checking CombineTypes/RecursiveRecordTypeMerge

### PR DESCRIPTION
**Pull request by FintanH**
Fixes #91 

he implementation brings with it intersection_with_key over HashMaps to
help with the type checking of records of records.

The implementation first checks that the Const values line up with the
LHS and RHS. Then checks that combining the records does not result in a
FieldCollision. It will finally return the shared Const type of the
arguments.

### CODE REVIEW

File: dhall/src/phase/typecheck.rs [hunk 1]
```diff
@@ -598,6 +598,83 @@ fn type_last_layer(
...
+                (dhall_syntax::Const::Kind, dhall_syntax::Const::Kind) => dhall_syntax::Const::Kind,
+                (dhall_syntax::Const::Sort, dhall_syntax::Const::Sort) => dhall_syntax::Const::Sort,
+                (l_mismatch, r_mismatch) =>
+                    return Err(mkerr(RecordTypeMismatch(
+                        Typed::from_const(l_mismatch),
+                        Typed::from_const(r_mismatch),
+                        l.clone(),
+                        r.clone(),
+                    ))),
+            };
```
Comment (Nadrieril) [inline dhall/src/phase/typecheck.rs]: I believe you could simply write `if k_l == k_r { k_l } else { return Err(...) }`

**Comment by Nadrieril** [comment]
You're on a roll !

File: dhall/src/phase/typecheck.rs (hunk 1 above)
Comment (FintanH) [reply dhall/src/phase/typecheck.rs]: Good call :+1:

**Comment by Nadrieril** [comment]
Just a conflict and then I can merge

**Comment by FintanH** [comment]
Hell of a conflict too! The `HashMap` functions end up being so similar that git thinks that some of the lines are shared :sweat_smile: Should be ready in a moment

**Comment by FintanH** [comment]
@Nadrieril I'm not sure how to discern the output of codecov :thinking:

**Comment by Nadrieril** [comment]
@FintanH It's fine, we don't care about losing coverage. Codecov is just useful info for when we're curious

**Comment by FintanH** [comment]
Alrighty, onward and upward!

**Pull request merged**